Envest/45 reduced gene sets

envest opened the pull request:
Closes #45 

Here we explore if using a targeted gene set with clinical applicability performs as well as the universal gene set, and how a targeted gene set performs next to a random gene set of the same size. We can do this experiment using kTSP, RF, and lasso models. MM2S and medulloPackage did not work in this context.

The main points of this PR:

- Train and test models using targeted and random gene sets in `predict/predict_targeted_gene_set.R`
- Compare kappa and accuracy values between the full, random, and targeted gene sets in `analysis_notebooks/targeted_gene_set.Rmd`
- Add the targeted gene set list in `processed_data/NS_IO_360_v1.0_Genes.tsv`
- Update `renv.lock` with `rstatix` and friends

Notes:

A previous version of this used the same training/testing study selection seed across the different gene sets. This allowed for pairwise comparison between gene sets that used the same studies and I calculated "delta kappa" values to show the difference in performance. However, I found that measure to be confusing, and also that using the same seed was an unnecessary constraint especially after the introduction of medulloPackage forced certain limitations on study selection. So... I just used different starting seeds to split training/testing, and now compare Kappa values directly instead of calculating a pairwise delta kappa. I think this also allows for more intuitive statistical comparison between the gene sets.

Questions for reviewer:

1. I used `renv::snapshot(type = "all")` to create a new `renv.lock` file since some of the existing packages in `renv.lock` were not explicitly referenced in our codebase. I wonder if that's advised, or if using the "implicit" default is better.
2. How best to run these scripts? Add an `06_targeted_gene_set.sh` for both model prediction and the analysis notebook?

Thank you!

jaclyn-taroni reviewed (commented):
Thanks for doing this! We should talk more about this at our meeting before you make any changes (_if_ you make changes). 

You've introduced additional variability by not using the same training/testing split. As in, it's harder to tell if there's a difference in performance because you're using targeted gene sets or because you happen to be testing on different datasets (probably not an issue for RNA-seq in practice and may not be an issue for arrays either). You could use the same splits and not delta Kappa as a measure – you could use a Wilcoxon test for paired samples instead. 

I also think that having different sample sizes for the full/targeted vs. random gene sets is a bit of an issue for display – a boxplot obscures that information, and (this is a personal problem) I'm reading too much into the differences in statistics. 

I agree that we're dealing with two questions here:

> 1. Given a targeted gene set of cancer-related genes, how does model performance compare between targeted gene set models and full gene set models?
> 2. Does the identity of the genes in the targeted gene set matter? In other words, can a random gene set of the same size perform just as well as the targeted gene set?

I think the first question is best answered through a paired test. I think the second question is best addressed via permutation testing, as I say in one of my comments. I'm finding one series of displays for both questions confusing.

Again, let's chat before you change anything!

jaclyn-taroni commented:
To answer your questions:

> 1. I used `renv::snapshot(type = "all")` to create a new `renv.lock` file since some of the existing packages in `renv.lock` were not explicitly referenced in our codebase. I wonder if that's advised, or if using the "implicit" default is better.

We have, so far, been handling this by adding `library({package in question})` to [`dependencies.R`](https://github.com/AlexsLemonade/medulloblastoma-classifier/blob/main/dependencies.R) and using the default for `renv::snapshot()`.

> 2. How best to run these scripts? Add an `06_targeted_gene_set.sh` for both model prediction and the analysis notebook?

That seems very reasonable to me 👍🏻

envest commented:
Thank you 👏 

Broad strokes comment ahead of our meeting in person tomorrow...
- I like your option c of keeping the test/train splits consistent but just comparing kappa, not delta kappa
  - Paired wilcoxon and effect size calculation will be :chefs_kiss:
- Agree that splitting the two questions up using a different comparison method makes sense
- Permutation testing is my favorite testing. I think the time to run many permutations will be fine. We can discuss more about how to structure this. e.g. For every test/train split run above, also do perm test of that targeted gene set vs. 1000+ random sets, so we get 10 repeated permutation tests with different train/test sets?
  - Just realized that `test set` is a palindrome 🤯 
- Will revert my `renv.lock` and redo the snapshot how you describe. Much cleaner.

envest commented on predict/predict_targeted_gene_set.R:
```diff
@@ -0,0 +1,332 @@
+  targeted_list <- purrr::map2(targeted_kTSP_RF_weighted_models_list,
+                               targeted_kTSP_RF_lasso_unweighted_models_list,
+                               c) |>
+    purrr::map(\(x) x[!duplicated(names(x))]) # remove duplicate list items
```
Each list contains some metadata that is the same, so this removes those redundant items.

envest commented on predict/predict_targeted_gene_set.R:
```diff
@@ -0,0 +1,332 @@
+# MM2S uses a pathway-based approach, and those pathways are pre-defined in `MM2S::HumanGMT$genesets`.
+# To return a GSVA score for a particular pathway, MM2S requires that between 20 and 100 genes overlap between the test data and the set of genes in that pathway.
+# MM2S later requires that there are at least 24 such pathways in the data.
+# However, with our targeted gene set, only n_mm2s_common_pathways pathways match the `20 <= overlapping genes <= 100` criteria.
```
I moved this to the analysis notebook and it gets printed out.

envest commented on analysis_notebooks/targeted_gene_set.Rmd:
````diff
@@ -0,0 +1,251 @@
+  dplyr::arrange(p.adj) |>
+  knitr::kable()
+
+```
````
This is now part of the analysis notebook!

envest commented on predict/predict_targeted_gene_set.R:
```diff
@@ -0,0 +1,332 @@
+random_gene_set_list <- purrr::map(1:10,
+                                   \(x) sample(row.names(bulk_genex_df),
+                                               size = n_targeted_genes_overlap,
+                                               replace = FALSE))
```
This got implemented using `tryCatch()` to allow for inexplicable random gene set model failures.

envest commented:
This is ready for the second round review! Key changes:

- Use the same training/test split when comparing full gene set vs. targeted gene set
- Use the same training/test split when comparing random gene set vs. targeted gene set
- Fully separate comparisons of full gene set vs. targeted gene set and random gene set vs. targeted gene set
- Use paired Wilcox test and effect size to compare full gene set vs. targeted gene set
- Use permutation test to compare random gene set vs. targeted gene set
- Reverted `renv.lock` and updated it the right way
- Added `06-targeted_gene_set.sh` to run modeling script and analysis notebook

Questions for reviewer:

- How to restructure/combine statistics and visualizations in analysis notebook
- `rstatix` didn't automatically do p-value correction as it has done previously when there were multiple comparisons. Add this back? p-value correction is covered in 🪄 Defense Against the Dark Arts 🪄 at Hogwarts.

Thanks!

envest commented on analysis_notebooks/targeted_gene_set.Rmd:
```diff
@@ -0,0 +1,385 @@
+                                                            "rf_weighted" ~ "RF (w)",
+                                                            "rf_unweighted" ~ "RF (unw)",
+                                                            "lasso" ~ "LASSO"))) |>
+  purrr::list_rbind(names_to = "random_index")
```
The names of the list elements from `purrr::map()` are 1:1000, and this sets the column `random_index` to be those numeric values. It's not clear to me when `purrr::map()` automatically sets list element names with the input but not this time!

envest commented:
More interpretation, better figures, and an explanation about medulloPackage not working... thank you for your months of discussion and reviews (and patience!)
